mesa/experimental/cell_space/grid.py

```diff
@@ -105,7 +106,103 @@ def _connect_single_cell_2d(self, cell: T, offsets: list[tuple[int, int]]) -> No
+    Methods:
+        add_property_layer(property_layer): Adds a new property layer to the grid.
+        remove_property_layer(property_name): Removes a property layer from the grid by its name.
+        select_random_empty_cell(self): Returns an empty random cell instance from the grid.
+
+
+    Note:
+        This class is not intended for direct use in user models but is currently used by the OrthogonalMooreGrid and OrthogonalVonNeumannGrid.
+    """
+
+    def __init__(
+        self,
+        dimensions: Sequence[int],
+        torus: bool = False,
+        capacity: float | None = None,
+        random: Random | None = None,
+        cell_klass: type[T] = Cell,
+        property_layers: None | PropertyLayer | list[PropertyLayer] = None,
+    ) -> None:
+        super().__init__(
+            dimensions=dimensions,
+            torus=torus,
+            capacity=capacity,
+            random=random,
+            cell_klass=cell_klass,
+        )
+
+        self.properties = {}
+
+        # Handle both single PropertyLayer instance and list of PropertyLayer instances
+        if property_layers:
+            # If a single PropertyLayer is passed, convert it to a list
+            if isinstance(property_layers, PropertyLayer):
+                property_layers = [property_layers]
+
+            for layer in property_layers:
+                self.add_property_layer(layer)
+
+    def add_property_layer(self, property_layer: PropertyLayer):
+        """
+        Adds a new property layer to the grid.
+
+        Args:
+            property_layer (PropertyLayer): The PropertyLayer instance to be added to the grid.
+
+        Raises:
+            ValueError: If a property layer with the same name already exists in the grid.
+            ValueError: If the dimensions of the property layer do not match the grid's dimensions.
+        """
+        if property_layer.name in self.properties:
+            raise ValueError(f"Property layer {property_layer.name} already exists.")
+        if (
+            property_layer.width != self.dimensions[0]
+            or property_layer.height != self.dimensions[1]
+        ):
+            raise ValueError(
+                f"Property layer dimensions {property_layer.width}x{property_layer.height} do not match grid dimensions {self.dimensions[0]}x{self.dimensions[1]}."
+            )
+        self.properties[property_layer.name] = property_layer
+
+    def remove_property_layer(self, property_name: str):
+        """
+        Removes a property layer from the grid by its name.
+
+        Args:
+            property_name (str): The name of the property layer to be removed.
+
+        Raises:
+            ValueError: If a property layer with the given name does not exist in the grid.
+        """
+        if property_name not in self.properties:
+            raise ValueError(f"Property layer {property_name} does not exist.")
+        del self.properties[property_name]
+
+    def select_random_empty_cell(self) -> T:
```

Comment:
Sorry to leave such a high-level comment so late in the review process, but I'm in doubt about the usage of a dedicated `select_random_empty_cell` method. In the original `_PropertyGrid` you would do this by calling `select_cells(only_empty=True)`. This is a more flexible approach, since you then can also add other conditions or masks.

While I'm not opposed to a PR that not integrates all the current functionality, I would like the methods that we add to be a (strict) subset of the current functionality, unless there's clear motivation to do other wise (performance could be one of them).

So, to move forward I would suggest one of three options:
- Extend the functionality to match or be a subset of [`select_cells()`](https://github.com/projectmesa/mesa/blob/8a50f9f36392b95824e0e2768dcf6f16515514d8/mesa/space.py#L889)
- Prove that there are clear benefits (performance or otherwise) of having a dedicated `select_random_empty_cell()` method.
- Remove the `select_random_empty_cell()` and merge this PR without it.

Reply:
Thank you for your feedback and for bringing up this concern. there isn't a clear performance benefit to having the select_random_empty_cell method. Its inclusion was primarily to ensure the property was added to the cell instance upon it's creation.

 I will proceed by removing the select_random_empty_cell method from the pull request.

Thank you for your guidance, and I apologize for any confusion this may have caused.

Reply:
Thanks! And no need to apologise at all. This is the normal process :)